zkir/Dialect/Field/IR/FieldTypes.td

```diff
@@ -101,13 +139,167 @@ def Field_QuadraticExtFieldType : Field_Type<"QuadraticExtField", "f2", [
     bool isMontgomery() const {
       return getBaseField().getIsMontgomery();
     }
+
+    // ExtensionFieldTypeInterface methods
+    unsigned getDegreeOverPrime() const {
+      return 2;
+    }
+
+    unsigned getDegreeOverBase() const;
+
+    ::mlir::Type getBaseFieldAsType() const {
+      return getBaseField();
+    }
+
+    ::mlir::Attribute getIrreducibleElementAsAttr();
+
+    ::mlir::Type cloneWithBaseFieldAndElement(::mlir::Type baseField,
+                                              ::mlir::Attribute element);
+
+    ::mlir::TypedAttr createConstantAttr(::llvm::ArrayRef<::llvm::APInt> coeffs);
+
+    ::mlir::Value buildStructFromCoeffs(::mlir::ImplicitLocOpBuilder &builder,
+                                        ::mlir::Type structType,
+                                        ::llvm::ArrayRef<::mlir::Value> coeffs) {
+      return ::mlir::zkir::SimpleStructBuilder<2>::initialized(
+          builder, builder.getLoc(), structType, coeffs);
+    }
+
+    ::llvm::SmallVector<::mlir::Value> extractCoeffsFromStruct(
+        ::mlir::ImplicitLocOpBuilder &builder, ::mlir::Value structValue) {
+      ::mlir::zkir::SimpleStructBuilder<2> extFieldStruct(structValue);
+      return extFieldStruct.getValues(builder, builder.getLoc());
+    }
+  }];
+  let extraClassDefinition = [{
+    unsigned $cppClass::getDegreeOverBase() const {
+      if (auto extBase = ::mlir::dyn_cast<ExtensionFieldTypeInterface>(
+              getBaseFieldAsType())) {
+        return getDegreeOverPrime() / extBase.getDegreeOverPrime();
+      }
+      return getDegreeOverPrime();
+    }
+
+    ::mlir::Attribute $cppClass::getIrreducibleElementAsAttr() {
+      return getBeta();
```

Comment:
Maybe we also need to name uniformly for extension field as non residue, otherwise we need to determine this name for every extension field..!